activerecord/lib/active_record/reflection.rb

```diff
@@ -236,14 +236,16 @@ def constraints
       end
 
       def counter_cache_column
-        @counter_cache_column ||= if belongs_to?
-          if options[:counter_cache] == true
-            -"#{active_record.name.demodulize.underscore.pluralize}_count"
-          elsif options[:counter_cache]
-            -options[:counter_cache].to_s
+        @counter_cache_column ||= begin
+          counter_cache = options[:counter_cache]
+
+          if belongs_to?
+            if counter_cache
+              counter_cache[:column] || -"#{active_record.name.demodulize.underscore.pluralize}_count"
+            end
+          else
+            (counter_cache && -counter_cache[:column]) || "#{name}_count"
```

Comment:
```suggestion
            (counter_cache && -counter_cache[:column]) || -"#{name}_count"
```

No big deal but we probably ought to intern this fallback too.